File: custom-plugins/cordova-plugin-wifi-billing-agent/src/ios/agents/BillingAgent.m
//
//  BillingAgent.m
//  WiFi Billing Agent
//

#import "BillingAgent.h"
#import "CoreAgent.h"

@interface BillingSession : NSObject
@property (nonatomic, strong) NSString *sessionId;
@property (nonatomic, strong) NSString *connectionId;
@property (nonatomic, strong) NSString *ssid;
@property (nonatomic, strong) NSString *userId;
@property (nonatomic, strong) NSString *planId;
@property (nonatomic, assign) NSTimeInterval startTime;
@property (nonatomic, assign) NSTimeInterval timeLimit;
@property (nonatomic, assign) long long dataLimit;
@property (nonatomic, assign) double rate;
@property (nonatomic, strong) NSString *state; // active, paused, ended
@property (nonatomic, assign) long long startBytes;
@property (nonatomic, assign) long long dataUsedBeforePause;
@property (nonatomic, assign) NSTimeInterval pauseTime;
@property (nonatomic, assign) NSTimeInterval totalPaused;
@property (nonatomic, assign) double finalCost;
@end

@implementation BillingSession
@end

@interface BillingAgent()

@property (nonatomic, strong) NSMutableDictionary<NSString *, BillingSession *> *activeSessions;
@property (nonatomic, strong) NSTimer *billingTimer;
@property (nonatomic, assign) NSTimeInterval billingInterval;
@property (nonatomic, assign) NSTimeInterval gracePeriod;

@end

@implementation BillingAgent

- (instancetype)initWithConfig:(NSDictionary *)config {
    self = [super init];
    if (self) {
        _currency = config[@"currency"] ?: @"USD";
        _defaultRate = [config[@"defaultRate"] doubleValue] ?: 0.10;
        _billingInterval = [config[@"billingInterval"] doubleValue] ?: 60.0;
        _gracePeriod = [config[@"gracePeriod"] doubleValue] ?: 30.0;
        _activeSessions = [[NSMutableDictionary alloc] init];
    }
    return self;
}

- (void)initialize {
    NSLog(@"[BillingAgent] Initializing");
    [self startBillingCycle];
}

- (void)shutdown {
    NSLog(@"[BillingAgent] Shutting down");
    [self.billingTimer invalidate];
    
    // End all active sessions
    for (NSString *sessionId in self.activeSessions.allKeys) {
        [self endSession:sessionId reason:@"agent_shutdown"];
    }
}

- (NSDictionary *)startSession:(NSString *)connectionId 
                          ssid:(NSString *)ssid 
                        userId:(NSString *)userId 
                        planId:(NSString *)planId 
                     timeLimit:(NSInteger)timeLimit 
                     dataLimit:(long long)dataLimit {
    
    NSString *sessionId = [NSString stringWithFormat:@"pc-sess-%.0f-%u", 
        [[NSDate date] timeIntervalSince1970] * 1000,
        arc4random_uniform(100000)];
    
    BillingSession *session = [[BillingSession alloc] init];
    session.sessionId = sessionId;
    session.connectionId = connectionId;
    session.ssid = ssid;
    session.userId = userId;
    session.planId = planId ?: @"default";
    session.startTime = [[NSDate date] timeIntervalSince1970] * 1000;
    session.timeLimit = timeLimit * 60 * 1000; // Convert to ms
    session.dataLimit = dataLimit;
    session.rate = self.defaultRate;
    session.state = @"active";
    session.startBytes = 0; // Would get from network interface
    session.totalPaused = 0;
    
    self.activeSessions[sessionId] = session;
    
    // Show local notification
    [self showSessionNotification:session];
    
    NSLog(@"[BillingAgent] Started session: %@ for SSID: %@", sessionId, ssid);
    
    return @{
        @"sessionId": sessionId,
        @"ssid": ssid,
        @"startTime": @(session.startTime),
        @"state": @"active",
        @"currency": self.currency,
        @"rate": @(session.rate)
    };
}

- (NSDictionary *)endSession:(NSString *)sessionId reason:(NSString *)reason {
    BillingSession *session = self.activeSessions[sessionId];
    if (!session) {
        return @{@"error": @"Session not found"};
    }
    
    // Calculate final usage
    NSTimeInterval elapsed = ([[NSDate date] timeIntervalSince1970] * 1000) - session.startTime - session.totalPaused;
    double cost = (elapsed / 1000.0 / 60.0) * session.rate;
    
    session.state = @"ended";
    session.finalCost = cost;
    
    // Clear notification
    [[NSNotificationCenter defaultCenter] postNotificationName:@"ClearBillingNotification" object:nil];
    
    NSDictionary *result = @{
        @"sessionId": sessionId,
        @"duration": @(elapsed / 1000 / 60),
        @"finalCost": @(cost),
        @"currency": self.currency,
        @"reason": reason
    };
    
    [self.activeSessions removeObjectForKey:sessionId];
    
    NSLog(@"[BillingAgent] Ended session: %@, Cost: $%.2f", sessionId, cost);
    
    return result;
}

- (NSDictionary *)pauseSession:(NSString *)sessionId reason:(NSString *)reason {
    BillingSession *session = self.activeSessions[sessionId];
    if (!session || ![session.state isEqualToString:@"active"]) {
        return @{@"error": @"Cannot pause inactive session"};
    }
    
    session.state = @"paused";
    session.pauseTime = [[NSDate date] timeIntervalSince1970] * 1000;
    session.dataUsedBeforePause = 0; // Would calculate from network stats
    
    return @{
        @"sessionId": sessionId,
        @"state": @"paused",
        @"reason": reason,
        @"pauseTime": @(session.pauseTime)
    };
}

- (NSDictionary *)resumeSession:(NSString *)sessionId {
    BillingSession *session = self.activeSessions[sessionId];
    if (!session || ![session.state isEqualToString:@"paused"]) {
        return @{@"error": @"Cannot resume non-paused session"};
    }
    
    NSTimeInterval pausedDuration = ([[NSDate date] timeIntervalSince1970] * 1000) - session.pauseTime;
    session.totalPaused += pausedDuration;
    session.state = @"active";
    
    return @{
        @"sessionId": sessionId,
        @"state": @"active",
        @"pausedDuration": @(pausedDuration)
    };
}

- (NSDictionary *)getSessionStatus:(NSString *)sessionId {
    BillingSession *session = self.activeSessions[sessionId];
    if (!session) {
        return @{@"active": @(NO)};
    }
    
    NSTimeInterval now = [[NSDate date] timeIntervalSince1970] * 1000;
    NSTimeInterval elapsed = now - session.startTime - session.totalPaused;
    NSTimeInterval remaining = session.timeLimit - elapsed;
    
    // Calculate data usage (simplified)
    long long dataUsed = 0; // Would get from network interface
    
    double currentCost = (elapsed / 1000.0 / 60.0) * session.rate;
    
    return @{
        @"active": @(YES),
        @"sessionId": sessionId,
        @"state": session.state,
        @"ssid": session.ssid,
        @"elapsedMinutes": @(elapsed / 1000 / 60),
        @"remainingMinutes": @(MAX(0, remaining / 1000 / 60)),
        @"dataUsed": @(dataUsed),
        @"dataUsedMB": @(dataUsed / (1024 * 1024)),
        @"currentCost": @(currentCost),
        @"currency": self.currency
    };
}

- (NSDictionary *)getDataUsage:(NSString *)sessionId {
    BillingSession *session = self.activeSessions[sessionId];
    if (!session) {
        return @{@"active": @(NO)};
    }
    
    long long dataUsed = 0; // Would calculate from network stats
    
    return @{
        @"active": @(YES),
        @"sessionDataUsed": @(dataUsed),
        @"sessionDataUsedMB": @(dataUsed / (1024 * 1024)),
        @"dataLimitMB": @(session.dataLimit / (1024 * 1024)),
        @"percentageUsed": session.dataLimit > 0 ? @((dataUsed * 100) / session.dataLimit) : @(0)
    };
}

- (void)startBillingCycle {
    self.billingTimer = [NSTimer scheduledTimerWithTimeInterval:self.billingInterval
                                                         target:self
                                                       selector:@selector(updateSessions)
                                                       userInfo:nil
                                                        repeats:YES];
}

- (void)updateSessions {
    NSTimeInterval now = [[NSDate date] timeIntervalSince1970] * 1000;
    
    for (BillingSession *session in self.activeSessions.allValues) {
        if (![session.state isEqualToString:@"active"]) continue;
        
        NSTimeInterval elapsed = now - session.startTime - session.totalPaused;
        NSTimeInterval remaining = session.timeLimit - elapsed;
        
        // Check time limit
        if (remaining <= 0) {
            [self handleTimeLimitReached:session];
        }
        
        // Check data limit
        if (session.dataLimit > 0 && 0 >= session.dataLimit) { // Simplified
            [self handleDataLimitReached:session];
        }
        
        // Low time warning
        if (remaining < 5 * 60 * 1000 && remaining > 4 * 60 * 1000) {
            [self.core emitEvent:@"billing:time_low" data:@{
                @"sessionId": session.sessionId,
                @"minutes": @(remaining / 1000 / 60)
            }];
        }
        
        // Update notification
        [self updateSessionNotification:session];
    }
}

- (void)handleTimeLimitReached:(BillingSession *)session {
    NSLog(@"[BillingAgent] Time limit reached for session: %@", session.sessionId);
    [self endSession:session.sessionId reason:@"time_limit_reached"];
    
    [self.core emitEvent:@"billing:time_limit_reached" data:@{
        @"sessionId": session.sessionId
    }];
}

- (void)handleDataLimitReached:(BillingSession *)session {
    NSLog(@"[BillingAgent] Data limit reached for session: %@", session.sessionId);
    [self endSession:session.sessionId reason:@"data_limit_reached"];
    
    [self.core emitEvent:@"billing:data_limit_reached" data:@{
        @"sessionId": session.sessionId
    }];
}

- (void)showSessionNotification:(BillingSession *)session {
    // Post notification for UI to show
    [[NSNotificationCenter defaultCenter] postNotificationName:@"ShowBillingNotification" 
                                                        object:nil 
                                                      userInfo:@{
                                                          @"title": @"WiFi Session Started",
                                                          @"body": [NSString stringWithFormat:@"Connected to %@", session.ssid]
                                                      }];
}

- (void)updateSessionNotification:(BillingSession *)session {
    NSTimeInterval now = [[NSDate date] timeIntervalSince1970] * 1000;
    NSTimeInterval elapsed = now - session.startTime - session.totalPaused;
    NSTimeInterval remaining = session.timeLimit - elapsed;
    double cost = (elapsed / 1000.0 / 60.0) * session.rate;
    
    [[NSNotificationCenter defaultCenter] postNotificationName:@"UpdateBillingNotification"
                                                        object:nil
                                                      userInfo:@{
                                                          @"title": [NSString stringWithFormat:@"WiFi Billing: %@", session.ssid],
                                                          @"body": [NSString stringWithFormat:@"Time: %.0f min | Cost: $%.2f", remaining / 1000 / 60, cost]
                                                      }];
}

@end

File: custom-plugins/cordova-plugin-aicore/cordova-plugin-wifi-billing-agent/src/android/agents/BillingAgent.java
package com.wifibilling.agent.agents;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.net.TrafficStats;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;

import androidx.core.app.NotificationCompat;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

public class BillingAgent {

    private static final String CHANNEL_ID = "WiFiBilling";
    private static final int NOTIFICATION_ID = 1001;

    private Context context;
    private CoreAgent coreAgent;
    private Handler mainHandler;
    private NotificationManager notificationManager;

    private Map<String, BillingSession> activeSessions = new HashMap<>();
    private Runnable billingRunnable;

    private String currency;
    private double defaultRate;
    private long billingInterval;

    public BillingAgent(Context context, CoreAgent coreAgent, JSONObject config) {
        this.context = context;
        this.coreAgent = coreAgent;
        this.mainHandler = new Handler(Looper.getMainLooper());
        this.notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);

        this.currency = config.optString("currency", "USD");
        this.defaultRate = config.optDouble("defaultRate", 0.10);
        this.billingInterval = config.optInt("billingInterval", 60000);

        createNotificationChannel();
    }

    public void initialize() {
        startBillingCycle();
    }

    public JSONObject startSession(String connectionId, String ssid, String userId,
            String planId, int timeLimit, long dataLimit) throws JSONException {

        String sessionId = "pc-sess-" + UUID.randomUUID().toString();

        BillingSession session = new BillingSession();
        session.id = sessionId;
        session.connectionId = connectionId;
        session.ssid = ssid;
        session.userId = userId;
        session.planId = planId;
        session.startTime = System.currentTimeMillis();
        session.timeLimit = timeLimit * 60 * 1000; // Convert to ms
        session.dataLimit = dataLimit;
        session.rate = defaultRate;
        session.state = "active";
        session.startBytes = TrafficStats.getTotalRxBytes() + TrafficStats.getTotalTxBytes();

        activeSessions.put(sessionId, session);

        // Show notification
        showSessionNotification(session);

        JSONObject result = new JSONObject();
        result.put("sessionId", sessionId);
        result.put("ssid", ssid);
        result.put("startTime", session.startTime);
        result.put("state", "active");
        result.put("currency", currency);
        result.put("rate", session.rate);

        return result;
    }

    public JSONObject endSession(String sessionId, String reason) throws JSONException {
        BillingSession session = activeSessions.get(sessionId);
        if (session == null) {
            throw new JSONException("Session not found: " + sessionId);
        }

        // Calculate final usage
        long currentBytes = TrafficStats.getTotalRxBytes() + TrafficStats.getTotalTxBytes();
        long dataUsed = currentBytes - session.startBytes;
        long duration = System.currentTimeMillis() - session.startTime;
        double cost = calculateCost(session, dataUsed, duration);

        session.endTime = System.currentTimeMillis();
        session.state = "ended";
        session.finalCost = cost;

        // Clear notification
        notificationManager.cancel(NOTIFICATION_ID);

        JSONObject result = new JSONObject();
        result.put("sessionId", sessionId);
        result.put("duration", duration / 1000 / 60); // minutes
        result.put("dataUsed", dataUsed);
        result.put("dataUsedMB", dataUsed / (1024 * 1024));
        result.put("finalCost", cost);
        result.put("currency", currency);
        result.put("reason", reason);

        activeSessions.remove(sessionId);

        return result;
    }

    public JSONObject getSessionStatus(String sessionId) throws JSONException {
        BillingSession session = activeSessions.get(sessionId);
        if (session == null) {
            JSONObject empty = new JSONObject();
            empty.put("active", false);
            return empty;
        }

        long currentBytes = TrafficStats.getTotalRxBytes() + TrafficStats.getTotalTxBytes();
        long dataUsed = currentBytes - session.startBytes;
        long elapsed = System.currentTimeMillis() - session.startTime;
        long remaining = session.timeLimit - elapsed;

        JSONObject status = new JSONObject();
        status.put("active", true);
        status.put("sessionId", sessionId);
        status.put("state", session.state);
        status.put("ssid", session.ssid);
        status.put("elapsedMinutes", elapsed / 1000 / 60);
        status.put("remainingMinutes", Math.max(0, remaining / 1000 / 60));
        status.put("dataUsed", dataUsed);
        status.put("dataUsedMB", dataUsed / (1024 * 1024));
        status.put("dataLimitMB", session.dataLimit / (1024 * 1024));
        status.put("currentCost", calculateCost(session, dataUsed, elapsed));
        status.put("totalCost", session.finalCost);
        status.put("currency", currency);

        return status;
    }

    public JSONObject pauseSession(String sessionId, String reason) throws JSONException {
        BillingSession session = activeSessions.get(sessionId);
        if (session == null)
            throw new JSONException("Session not found");

        session.state = "paused";
        session.pauseTime = System.currentTimeMillis();

        // Save current data usage
        long currentBytes = TrafficStats.getTotalRxBytes() + TrafficStats.getTotalTxBytes();
        session.dataUsedBeforePause = currentBytes - session.startBytes;

        JSONObject result = new JSONObject();
        result.put("sessionId", sessionId);
        result.put("state", "paused");
        result.put("reason", reason);
        result.put("pauseTime", session.pauseTime);

        return result;
    }

    public JSONObject resumeSession(String sessionId) throws JSONException {
        BillingSession session = activeSessions.get(sessionId);
        if (session == null)
            throw new JSONException("Session not found");

        long pausedDuration = System.currentTimeMillis() - session.pauseTime;
        session.totalPaused += pausedDuration;
        session.state = "active";

        // Reset byte counters
        session.startBytes = TrafficStats.getTotalRxBytes() + TrafficStats.getTotalTxBytes()
                - session.dataUsedBeforePause;

        JSONObject result = new JSONObject();
        result.put("sessionId", sessionId);
        result.put("state", "active");
        result.put("pausedDuration", pausedDuration);

        return result;
    }

    public JSONObject getDataUsage(String sessionId) throws JSONException {
        BillingSession session = activeSessions.get(sessionId);
        if (session == null) {
            JSONObject empty = new JSONObject();
            empty.put("active", false);
            return empty;
        }

        long currentBytes = TrafficStats.getTotalRxBytes() + TrafficStats.getTotalTxBytes();
        long dataUsed = currentBytes - session.startBytes;

        JSONObject usage = new JSONObject();
        usage.put("active", true);
        usage.put("sessionDataUsed", dataUsed);
        usage.put("sessionDataUsedMB", dataUsed / (1024 * 1024));
        usage.put("dataLimitMB", session.dataLimit / (1024 * 1024));
        usage.put("percentageUsed", session.dataLimit > 0 ? (dataUsed * 100 / session.dataLimit) : 0);

        return usage;
    }

    private void startBillingCycle() {
        billingRunnable = new Runnable() {
            @Override
            public void run() {
                for (BillingSession session : activeSessions.values()) {
                    if ("active".equals(session.state)) {
                        try {
                            updateSession(session);
                        } catch (Exception e) {
                            e.printStackTrace();
                        }
                    }
                }
                mainHandler.postDelayed(this, billingInterval);
            }
        };
        mainHandler.postDelayed(billingRunnable, billingInterval);
    }

    private void updateSession(BillingSession session) throws JSONException {
        long currentBytes = TrafficStats.getTotalRxBytes() + TrafficStats.getTotalTxBytes();
        long dataUsed = currentBytes - session.startBytes;
        long elapsed = System.currentTimeMillis() - session.startTime - session.totalPaused;

        // Check time limit
        if (elapsed >= session.timeLimit) {
            // Trigger time limit event
            coreAgent.emitEvent("billing:time_limit_reached",
                    new JSONObject().put("sessionId", session.id));
        }

        // Check data limit
        if (session.dataLimit > 0 && dataUsed >= session.dataLimit) {
            coreAgent.emitEvent("billing:data_limit_reached",
                    new JSONObject().put("sessionId", session.id));
        }

        // Low time warning
        long remaining = session.timeLimit - elapsed;
        if (remaining < 5 * 60 * 1000 && remaining > 4 * 60 * 1000) { // 5 minute warning
            coreAgent.emitEvent("billing:time_low",
                    new JSONObject()
                            .put("sessionId", session.id)
                            .put("minutes", remaining / 1000 / 60));
        }

        // Update notification
        updateSessionNotification(session, elapsed, dataUsed);
    }

    private double calculateCost(BillingSession session, long dataUsed, long duration) {
        // Time-based billing
        double timeCost = (duration / 1000.0 / 60.0) * session.rate;

        // Could add data-based component here
        // double dataCost = (dataUsed / (1024.0 * 1024.0 * 1024.0)) * dataRate;

        return Math.round(timeCost * 100.0) / 100.0;
    }

    private void createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel = new NotificationChannel(
                    CHANNEL_ID,
                    "WiFi Billing Session",
                    NotificationManager.IMPORTANCE_LOW);
            channel.setDescription("Active WiFi billing session");
            notificationManager.createNotificationChannel(channel);
        }
    }

    private void showSessionNotification(BillingSession session) {
        Notification notification = buildNotification("Session started",
                "Connected to " + session.ssid);
        notificationManager.notify(NOTIFICATION_ID, notification);
    }

    private void updateSessionNotification(BillingSession session, long elapsed, long dataUsed) {
        long remainingMinutes = Math.max(0, (session.timeLimit - elapsed) / 1000 / 60);
        double currentCost = calculateCost(session, dataUsed, elapsed);

        String content = String.format("Time: %d min remaining | Cost: $%.2f",
                remainingMinutes, currentCost);

        Notification notification = buildNotification("WiFi Billing Active", content);
        notificationManager.notify(NOTIFICATION_ID, notification);
    }

    private Notification buildNotification(String title, String content) {
        NotificationCompat.Builder builder = new NotificationCompat.Builder(context, CHANNEL_ID)
                .setContentTitle(title)
                .setContentText(content)
                .setSmallIcon(android.R.drawable.ic_menu_info_details)
                .setOngoing(true)
                .setPriority(NotificationCompat.PRIORITY_LOW);

        return builder.build();
    }

    public void shutdown() {
        mainHandler.removeCallbacks(billingRunnable);
        notificationManager.cancel(NOTIFICATION_ID);
    }

    // Inner class for session tracking
    private static class BillingSession {
        String id;
        String connectionId;
        String ssid;
        String userId;
        String planId;
        long startTime;
        long endTime;
        long timeLimit;
        long dataLimit;
        long startBytes;
        long dataUsedBeforePause;
        long pauseTime;
        long totalPaused;
        double rate;
        double finalCost;
        String state; // active, paused, ended
    }
}
